src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionResolveHandler.cs

```diff
@@ -128,6 +129,16 @@ public async Task<CompletionItem> HandleRequestAsync(CompletionItem request, Cli
             var formattingOptions = _formattingOptionsProvider.GetOptions(requestContext.HostDocumentUri);
```

Comment:
> * The core problem here is that C# isn't respecting the user's tabs/spaces setting when generating override completion TextEdits.

Hmm, super interesting problem. In an ideal world we'd have a way to update the C# buffer / Roslyn document to reflect the top-level Razor documents formatting options and then C# would respect those settings similarly to how we do that here. How does C# typically get their formatting options when generating override completion snippets?

Reply:
> C# and Razor can have different tabs/spaces settings

🤯 

wat

Reply:
> It looks like C# uses the document's formatting options to generate override completion snippets.

Hmmm, this definitely seems like a design flaw we have among our worlds. Given we populate the Razor file in the C# workspace is there a document service we can provide that modifies the C# settings to match Razor's?

Reply:
> 🤯
wat

😁 yeah, so to clarify, the tabs/spaces setting we use for all portions of a Razor document (incl. C# portions) is taken from the Razor Tools->Options setting. The C# Tools->Options option exists but isn't used for Razor files, which makes things super confusing because the C# LS now needs to know about Razor options when doing anything related to formatting in a Razor doc.

> Given we populate the Razor file in the C# workspace is there a document service we can provide that modifies the C# settings to match Razor's?

Hmm, I'm not aware of a document service today that does this, @dibarbet would you know?

Reply:
There is not currently a document service that does this.  Introducing that kind of service I think could be a good approach, but maybe we should also consider having options be a part of the dynamic file info API.  

We do use the the [IIndentationManagerService](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.text.editor.iindentationmanagerservice?view=visualstudiosdk-2019) from the editor to get the final settings, but I'm not certain we can map the dynamic files to a text buffer.

Reply:
Tagging @jasonmalinowski as well for thoughts on options + dynamic file info (and as the person who worked on IIndentationManagerService)

Reply:
So a few ways to look at this:

1. When we discussed dynamic files in the past, I think one general idea was that the document produced in our workspace should have a .cshtml file name; that way the formatting settings coming from .editorconfig that apply to a .cshtml would be picked up naturally. Is that not happening here? Should it?
2. Rather than hacking around it, should we just do the LSP work? I'm not entirely sure this is urgent in that sense.

Reply:
> C# and Razor can have different tabs/spaces settings

This has been a bug farm for ages. For IVsContainedLanguage cases there's [an API](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.textmanager.interop.ivscontainedlanguagehost.getlineindent) to get indenting information which we then apply in non-LSP cases.

Reply:
>Rather than hacking around it, should we just do the LSP work? I'm not entirely sure this is urgent in that sense.

I agree with this for indentation.  But. a question for the Razor folks: are there any options that 
1.  would not be client managed (and not come from the LSP request)
2.  differ between C# and Razor C#
3.  and also are not in editor config?

If there aren't any that meet this criteria then I think a combination of LSP passing the right client settings + editorconfig should resolve most of the cases (and not need to introduce anything to the dynamic file info / document service)